packages/ses/src/module-proxy.js

```diff
@@ -0,0 +1,173 @@
+const q = JSON.stringify;
+
+// `createDeferredModule` creates a module and activator.
+// A `Compartment` can create a module for any
+// module specifier, regardless of whether it is loadable or executable, and
+// use that object as a token that can be fed into another compartment's module
+// map.
+// Only after the specified module has been analyzed is it possible for the
+// module namespace proxy to behave properly, so it throws exceptions until
+// after the compartment has executed the module.
+export const createDeferredModule = () => {
+  let internalModule;
+  const target = create(null);
+  return freeze({
+    activate(module) {
+      internalModule = module;
+      defineProperties(target, getOwnPropertyDescriptors(module));
+      if (!isExtensible(module)) {
+        preventExtensions(target);
+      }
+      if (isSealed(module)) {
+        seal(target);
+      }
+    },
+    module: new Proxy(target, {
+      get(_target, name, receiver) {
+        if (internalModule === undefined) {
+          throw new TypeError(
+            `Cannot get property ${q(
+              name,
+            )} of module exports namespace, the module has not yet begun executing`,
+          );
+        }
+        return Reflect.get(internalModule, name, receiver);
+      },
+      set(_target, name, value) {
+        if (internalModule === undefined) {
+          throw new TypeError(
+            `Cannot set property ${q(
+              name,
+            )} of module exports namespace, the module has not yet begun executing`,
+          );
+        }
+        if (!Reflect.set(internalModule, name, value)) {
+          throw new TypeError(
+            `Cannot add property ${q(
+              name,
+            )}, the module exports namespace is not extensible`,
+          );
+        }
+        return true;
+      },
+      has(_target, name) {
+        if (internalModule === undefined) {
+          throw new TypeError(
+            `Cannot check property ${q(
+              name,
+            )}, the module has not yet begun to execute`,
+          );
+        }
+        return Reflect.has(target, name);
+      },
+      deleteProperty(_target, name) {
+        throw new TypeError(
+          `Cannot delete property ${q(
+            name,
+          )}, the module exports namespace is not extensible`,
+        );
+      },
+      ownKeys(_target) {
+        if (internalModule === undefined) {
+          throw new TypeError(
+            `Cannot enumerate keys, the module has not yet begun to execute`,
+          );
+        }
+        return Reflect.ownKeys(internalModule);
+      },
+      getOwnPropertyDescriptor(_target, name) {
+        if (internalModule === undefined) {
+          throw new TypeError(
+            `Cannot get own property descriptor ${q(
+              name,
+            )}, the module has not yet begun to execute`,
+          );
+        }
+        return Reflect.getOwnPropertyDescriptor(internalModule, name);
```

Comment:
This one is interesting. This descriptor will be checked against the `target`, which was initialized to the same descriptor state. But if the descriptor is writable, the value might change. Does the enforcement mechanism, in this case, ignore the target's property's value? (I think it should. I just don't remember if it does.)

Reply:
This does not seem right to me.

```js
let t = {};
let o = {a: 10};
Object.defineProperties(t, Object.getOwnPropertyDescriptors(o));
let p = new Proxy(t, {
  getOwnPropertyDescriptor(_t, name) {
    return Reflect.getOwnPropertyDescriptor(o, name);
  }
});
t.a = 30;
o.a = 20;
console.log(p.a);
// 30
```

Reply:
But I was looking at the wrong thing. This seems fine:

```js
let t = {};
let o = {a: 10};
Object.defineProperties(t, Object.getOwnPropertyDescriptors(o));
let p = new Proxy(t, {
  getOwnPropertyDescriptor(_t, name) {
    console.log('here');
    return Reflect.getOwnPropertyDescriptor(o, name);
  }
});
t.a = 30;
o.a = 20;
console.log(Object.getOwnPropertyDescriptor(p, 'a').value);
// 20
```